### PR TITLE
Api Key Support

### PR DESCRIPTION
**Pull request by Sushisource**
## What was changed
Add api-key helper methods for both kinds of clients

## Why?
Parity

## Checklist

1. Closes https://github.com/temporalio/sdk-typescript/issues/1367

2. How was this tested:
Upgraded tests

3. Any docs updates needed?

### CODE REVIEW

File: packages/client/src/connection.ts [hunk 1]
```diff
@@ -451,7 +466,39 @@ export class Connection {
...
+   *   conn.withMetadata({ otherKey: 'set' }, () => client.start(options)))
+   * );
+   * ```
+   */
+  async withApiKey<ReturnType>(apiKey: string, fn: () => Promise<ReturnType>): Promise<ReturnType> {
+    const cc = this.callContextStorage.getStore();
+    return await this.callContextStorage.run(
+      {
+        ...cc,
+        metadata: { Authorization: `Bearer ${apiKey}`, ...cc?.metadata },
```
[the remark this answers is not in the document]
Comment (Sushisource) [reply packages/client/src/connection.ts:498]: It works that way, though? You use this method to change it for any particular call.

[the remark this answers is not in the document]
Comment (cretz) [reply packages/client/src/connection.ts:498]: Hrmm, I might have expected that we'd mutate something the user can read back. But now I see unfortunately metadata wasn't even done that way. So there is no way for the user to retrieve the set API key or metadata? If I had a client and I wanted to remove a piece of metadata, how? Or if I wanted to check that I had an API key set, how would I? That's what I meant be "allows people to get and update API key independently of metadata", but it appears there is no "get" part in TS anyways?

Also, it is a bit unfortunate that for metadata (and now api key) we didn't make worker native connection and TS connection more similar in the ways of getting/setting these values.

But I guess these are general TypeScript SDK issues and not specific to this property, but if/when we ever offer the ability to get the API key or metadata, we will probably need to split these for easy use (get-then-update metadata is a common use case).

File: packages/client/src/connection.ts [hunk 2]
```diff
@@ -142,9 +150,16 @@ function addDefaults(options: ConnectionOptions): ConnectionOptionsWithDefaults
  * - Convert {@link ConnectionOptions.tls} to {@link grpc.ChannelCredentials}
  * - Add the grpc.ssl_target_name_override GRPC {@link ConnectionOptions.channelArgs | channel arg}
  * - Add default port to address if port not specified
+ * - Set `Authorization` header based on API key (and not already set explicitly)
  */
 function normalizeGRPCConfig(options?: ConnectionOptions): ConnectionOptions {
   const { tls: tlsFromConfig, credentials, callCredentials, ...rest } = options || {};
+  if (rest.apiKey && !rest.metadata?.['Authorization']) {
```
Comment (mjameswh) [inline packages/client/src/connection.ts:157]: I'd prefer we deal with this incoheencies with a TypeError, just like we do for the tls+credentials case below.

File: packages/client/src/connection.ts [hunk 3]
```diff
@@ -142,9 +150,16 @@ function addDefaults(options: ConnectionOptions): ConnectionOptionsWithDefaults
...
+    if (!rest.metadata) {
+      rest.metadata = {};
+    }
```
Comment (mjameswh) [inline packages/client/src/connection.ts:160]:
```
rest.metadata ??= {};
```

File: packages/client/src/connection.ts [hunk 4]
```diff
@@ -142,9 +150,16 @@ function addDefaults(options: ConnectionOptions): ConnectionOptionsWithDefaults
  * - Convert {@link ConnectionOptions.tls} to {@link grpc.ChannelCredentials}
  * - Add the grpc.ssl_target_name_override GRPC {@link ConnectionOptions.channelArgs | channel arg}
  * - Add default port to address if port not specified
+ * - Set `Authorization` header based on API key (and not already set explicitly)
  */
 function normalizeGRPCConfig(options?: ConnectionOptions): ConnectionOptions {
   const { tls: tlsFromConfig, credentials, callCredentials, ...rest } = options || {};
```
Comment (mjameswh) [inline packages/client/src/connection.ts:156]:
```suggestion
  const { tls: tlsFromConfig, credentials, callCredentials, metadata = {}, ...rest } = options || {};
```

... and then, replace `rest.metadata` everywhere below. No more need to check that metadata isn't empty.

File: packages/client/src/connection.ts (hunk 2 above)
Comment (mjameswh) [reply packages/client/src/connection.ts:157]: We should also forbid the credentials + apiKey use case, as `credentials` will override `metadata['Authorization']`

File: packages/client/src/connection.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (mjameswh) [reply packages/client/src/connection.ts:498]: Important! This must be:

```suggestion
        metadata: { ...cc?.metadata, Authorization: `Bearer ${apiKey}` },
```

In object spread, later entries override earlier ones, so the original version would not work.

File: packages/test/src/test-native-connection-headers.ts [hunk 5]
```diff
@@ -47,9 +51,14 @@ test('NativeConnection passes headers provided in options', async (t) => {
       >,
       callback: grpc.sendUnaryData<temporal.api.workflowservice.v1.IPollActivityTaskQueueResponse>
     ) {
+      console.log(call.metadata);
```
Comment (mjameswh) [inline packages/test/src/test-native-connection-headers.ts:54]: Leftover?

Comment (Sushisource) [reply packages/test/src/test-native-connection-headers.ts:54]: Yes, thanks

File: packages/client/src/connection.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Sushisource) [reply packages/client/src/connection.ts:498]: @mjameswh That was actually intentional, as the docstrings and the behavior I copied from the Python PR is that any `Authorization` header set in the metadata takes precedence.

File: packages/client/src/connection.ts [hunk 6]
```diff
@@ -142,9 +150,22 @@ function addDefaults(options: ConnectionOptions): ConnectionOptionsWithDefaults
...
+  if (rest.apiKey) {
+    if (rest.metadata?.['Authorization']) {
+      throw new TypeError(
+        'Both `apiKey` option and `Authorization` header were provided. Only one makes sense to use at a time.'
+      );
+    }
+    if (credentials !== undefined) {
+      throw new TypeError(
+        'Both `apiKey` and `credentials` ConnectionOptions were provided. Only one makes sense to use at a time.'
```
Comment (bergundy) [inline packages/client/src/connection.ts:165]: nit: I don't believe we typically put periods in our error messages.

File: packages/client/src/connection.ts [hunk 7]
```diff
@@ -100,6 +100,14 @@ export interface ConnectionOptions {
    */
   metadata?: Metadata;
 
+  /**
+   * API key for Temporal. This becomes the "Authorization" HTTP header with "Bearer " prepended.
+   * This is only set if RPC metadata doesn't already have an "authorization" key.
+   *
+   * In order to dynamically set metadata, use {@link Connection.withApiKey} or {@link Connection.setApiKey}
+   */
+  apiKey?: string;
```
Comment (bergundy) [inline packages/client/src/connection.ts:109]: If the setter accepts a function, you should accept a function here too:

```suggestion
  apiKey?: string | (() => string);
```

File: packages/client/src/connection.ts [hunk 8]
```diff
@@ -100,6 +100,14 @@ export interface ConnectionOptions {
    */
   metadata?: Metadata;
 
+  /**
+   * API key for Temporal. This becomes the "Authorization" HTTP header with "Bearer " prepended.
+   * This is only set if RPC metadata doesn't already have an "authorization" key.
```
Comment (mjameswh) [inline packages/client/src/connection.ts:105]: > This is only set if RPC metadata doesn't already have an "authorization" key.

This is no longer true. `apiKey` and `authorization` header are mutually exclusive. Also add a mention for this on `metadata`'s doc.

File: packages/client/src/connection.ts [hunk 9]
```diff
@@ -100,6 +100,14 @@ export interface ConnectionOptions {
...
+   *
+   * In order to dynamically set metadata, use {@link Connection.withApiKey} or {@link Connection.setApiKey}
```
Comment (mjameswh) [inline packages/client/src/connection.ts:107]:
```suggestion
   * In order to replace the API key at a later time, use {@link Connection.withApiKey} or {@link Connection.setApiKey}
```

File: packages/client/src/connection.ts (hunk 4 above)
Comment (mjameswh) [reply packages/client/src/connection.ts:156]: Yeah, forgot to mention that you would have to add `metadata` in object spread later on. That's still more JS-idiomatic.

File: packages/worker/src/connection-options.ts [hunk 10]
```diff
@@ -28,11 +28,18 @@ export interface NativeConnectionOptions {
    * Set statically at connection time, can be replaced later using {@link NativeConnection.setMetadata}.
    */
   metadata?: Record<string, string>;
+
+  /**
+   * API key for Temporal. This becomes the "Authorization" HTTP header with "Bearer " prepended.
+   * This is only set if RPC metadata doesn't already have an "authorization" key.
```
Comment (mjameswh) [inline packages/worker/src/connection-options.ts:34]: > This is only set if RPC metadata doesn't already have an "authorization" key.

On creation of a `NativeConntection`, we should check for apikey+authorization metadata, and throw just like in client connection.

File: packages/client/src/connection.ts (hunk 7 above)
Comment (Sushisource) [reply packages/client/src/connection.ts:109]: Sure, I considered it but thought they'd probably just set it after construction if they need it dynamic, but, why not.

File: packages/client/src/connection.ts (hunk 6 above)
Comment (Sushisource) [reply packages/client/src/connection.ts:165]: Changed, but I have no idea why this matters lol.

File: packages/client/src/connection.ts [hunk 11]
```diff
@@ -451,7 +497,52 @@ export class Connection {
...
+      },
+      fn
+    );
+  }
+
+  /**
+   * Set the {@link ConnectionOptions.apiKey} for all subsequent requests. A static string or a
+   * callback function may be provided.
+   */
+  setApiKey(apiKey: string | (() => string)): void {
```
Comment (bergundy) [inline packages/client/src/connection.ts:537]: I know we discussed on slack that we want this URL but if we also have a callback on creation can we maybe punt on adding this API until we get user feedback that the callback approach isn't good enough?

It's easy to add APIs, hard to remove them.

**Review by bergundy — APPROVED** [review]
Overall LGTM.

Should we also mark these APIs as experimental? What did we do in other SDKs?

File: packages/client/src/connection.ts (hunk 11 above)
Comment (Sushisource) [reply packages/client/src/connection.ts:537]: Sorry, URL?

**Pull request merged**